llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Fusion.cpp

```diff
@@ -19,20 +19,34 @@
 using namespace mlir;
 using namespace mlir::iree_compiler::IREE::LinalgExt;
 
-FailureOr<SmallVector<TilingInterface>>
+FailureOr<SmallVector<Operation *>>
 mlir::iree_compiler::IREE::LinalgExt::LinalgExtFusionInContainingOpPattern::
-    returningMatchAndRewrite(TilingInterface producerOp,
+    returningMatchAndRewrite(Operation *producerOp,
                              PatternRewriter &rewriter) const {
   if (producerOp->getNumResults() != 1)
     return failure();
 
+  auto tiled = tileAndFuse(producerOp, rewriter);
+  if (succeeded(tiled))
+    return tiled;
+
+  return cloneAndFuse(producerOp, rewriter);
+}
+
+FailureOr<SmallVector<Operation *>>
+mlir::iree_compiler::IREE::LinalgExt::LinalgExtFusionInContainingOpPattern::
+    tileAndFuse(Operation *producerOp, RewriterBase &rewriter) const {
+  auto tileableProducer = dyn_cast<TilingInterface>(producerOp);
+  if (!tileableProducer)
+    return failure();
+
   // Search the producer slices accessed within the containing operation.
   SmallVector<tensor::ExtractSliceOp> sliceOps;
-  for (Operation *user : producerOp->getUsers()) {
+  for (Operation *user : tileableProducer->getUsers()) {
     auto sliceOp = dyn_cast<tensor::ExtractSliceOp>(user);
     if (!sliceOp)
       continue;
-    if (sliceOp->getParentOp() != containingOp)
+    if (!containingOp->isProperAncestor(sliceOp))
```

Comment:
This relaxation would allow many loops and other containing ops to exist above the slice, is this what we want / well tested ?

Reply:
I don't have a particular use case in mind, but conceptually we only care that the slice is somewhere within the `scf.foreach_thread` op. Doesn't matter where, the transformation is the same.